// 0x.js/lib/src/utils/order_validation_utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash");
var types_1 = require("../types");
var _0x_1 = require("../0x");
var utils_1 = require("../utils/utils");
var constants_1 = require("../utils/constants");
var OrderValidationUtils = (function () {
    function OrderValidationUtils(tokenWrapper, exchangeWrapper) {
        this.tokenWrapper = tokenWrapper;
        this.exchangeWrapper = exchangeWrapper;
    }
    OrderValidationUtils.prototype.validateOrderFillableOrThrowAsync = function (signedOrder, zrxTokenAddress, expectedFillTakerTokenAmount) {
        return __awaiter(this, void 0, void 0, function () {
            var orderHash, unavailableTakerTokenAmount, fillTakerTokenAmount;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        orderHash = utils_1.utils.getOrderHashHex(signedOrder);
                        return [4 /*yield*/, this.exchangeWrapper.getUnavailableTakerAmountAsync(orderHash)];
                    case 1:
                        unavailableTakerTokenAmount = _a.sent();
                        this.validateRemainingFillAmountNotZeroOrThrow(signedOrder.takerTokenAmount, unavailableTakerTokenAmount);
                        this.validateOrderNotExpiredOrThrow(signedOrder.expirationUnixTimestampSec);
                        fillTakerTokenAmount = signedOrder.takerTokenAmount.minus(unavailableTakerTokenAmount);
                        if (!_.isUndefined(expectedFillTakerTokenAmount)) {
                            fillTakerTokenAmount = expectedFillTakerTokenAmount;
                        }
                        return [4 /*yield*/, this.validateFillOrderMakerBalancesAllowancesThrowIfInvalidAsync(signedOrder, fillTakerTokenAmount, zrxTokenAddress)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    OrderValidationUtils.prototype.validateFillOrderThrowIfInvalidAsync = function (signedOrder, fillTakerTokenAmount, takerAddress, zrxTokenAddress) {
        return __awaiter(this, void 0, void 0, function () {
            var orderHash, unavailableTakerTokenAmount, wouldRoundingErrorOccur;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (fillTakerTokenAmount.eq(0)) {
                            throw new Error(types_1.ExchangeContractErrs.OrderFillAmountZero);
                        }
                        orderHash = utils_1.utils.getOrderHashHex(signedOrder);
                        if (!_0x_1.ZeroEx.isValidSignature(orderHash, signedOrder.ecSignature, signedOrder.maker)) {
                            throw new Error(types_1.ZeroExError.InvalidSignature);
                        }
                        return [4 /*yield*/, this.exchangeWrapper.getUnavailableTakerAmountAsync(orderHash)];
                    case 1:
                        unavailableTakerTokenAmount = _a.sent();
                        this.validateRemainingFillAmountNotZeroOrThrow(signedOrder.takerTokenAmount, unavailableTakerTokenAmount);
                        if (signedOrder.taker !== constants_1.constants.NULL_ADDRESS && signedOrder.taker !== takerAddress) {
                            throw new Error(types_1.ExchangeContractErrs.TransactionSenderIsNotFillOrderTaker);
                        }
                        this.validateOrderNotExpiredOrThrow(signedOrder.expirationUnixTimestampSec);
                        return [4 /*yield*/, this.validateFillOrderBalancesAllowancesThrowIfInvalidAsync(signedOrder, fillTakerTokenAmount, takerAddress, zrxTokenAddress)];
                    case 2:
                        _a.sent();
                        return [4 /*yield*/, this.exchangeWrapper.isRoundingErrorAsync(fillTakerTokenAmount, signedOrder.takerTokenAmount, signedOrder.makerTokenAmount)];
                    case 3:
                        wouldRoundingErrorOccur = _a.sent();
                        if (wouldRoundingErrorOccur) {
                            throw new Error(types_1.ExchangeContractErrs.OrderFillRoundingError);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    OrderValidationUtils.prototype.validateFillOrKillOrderThrowIfInvalidAsync = function (signedOrder, fillTakerTokenAmount, takerAddress, zrxTokenAddress) {
        return __awaiter(this, void 0, void 0, function () {
            var orderHashHex, unavailableTakerAmount, remainingTakerAmount;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.validateFillOrderThrowIfInvalidAsync(signedOrder, fillTakerTokenAmount, takerAddress, zrxTokenAddress)];
                    case 1:
                        _a.sent();
                        orderHashHex = utils_1.utils.getOrderHashHex(signedOrder);
                        return [4 /*yield*/, this.exchangeWrapper.getUnavailableTakerAmountAsync(orderHashHex)];
                    case 2:
                        unavailableTakerAmount = _a.sent();
                        remainingTakerAmount = signedOrder.takerTokenAmount.minus(unavailableTakerAmount);
                        if (remainingTakerAmount < fillTakerTokenAmount) {
                            throw new Error(types_1.ExchangeContractErrs.InsufficientRemainingFillAmount);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    OrderValidationUtils.prototype.validateCancelOrderThrowIfInvalidAsync = function (order, cancelTakerTokenAmount, unavailableTakerTokenAmount) {
        return __awaiter(this, void 0, void 0, function () {
            var currentUnixTimestampSec;
            return __generator(this, function (_a) {
                if (cancelTakerTokenAmount.eq(0)) {
                    throw new Error(types_1.ExchangeContractErrs.OrderCancelAmountZero);
                }
                if (order.takerTokenAmount.eq(unavailableTakerTokenAmount)) {
                    throw new Error(types_1.ExchangeContractErrs.OrderAlreadyCancelledOrFilled);
                }
                currentUnixTimestampSec = utils_1.utils.getCurrentUnixTimestamp();
                if (order.expirationUnixTimestampSec.lessThan(currentUnixTimestampSec)) {
                    throw new Error(types_1.ExchangeContractErrs.OrderCancelExpired);
                }
                return [2 /*return*/];
            });
        });
    };
    OrderValidationUtils.prototype.validateFillOrderBalancesAllowancesThrowIfInvalidAsync = function (signedOrder, fillTakerAmount, senderAddress, zrxTokenAddress) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.validateFillOrderMakerBalancesAllowancesThrowIfInvalidAsync(signedOrder, fillTakerAmount, zrxTokenAddress)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.validateFillOrderTakerBalancesAllowancesThrowIfInvalidAsync(signedOrder, fillTakerAmount, senderAddress, zrxTokenAddress)];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    OrderValidationUtils.prototype.validateFillOrderMakerBalancesAllowancesThrowIfInvalidAsync = function (signedOrder, fillTakerAmount, zrxTokenAddress) {
        return __awaiter(this, void 0, void 0, function () {
            var makerBalance, makerAllowance, isMakerTokenZRX, exchangeRate, fillMakerAmount, requiredMakerAmount, makerZRXBalance, isTakerTokenZRX, makerZRXAllowance;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.tokenWrapper.getBalanceAsync(signedOrder.makerTokenAddress, signedOrder.maker)];
                    case 1:
                        makerBalance = _a.sent();
                        return [4 /*yield*/, this.tokenWrapper.getProxyAllowanceAsync(signedOrder.makerTokenAddress, signedOrder.maker)];
                    case 2:
                        makerAllowance = _a.sent();
                        isMakerTokenZRX = signedOrder.makerTokenAddress === zrxTokenAddress;
                        exchangeRate = signedOrder.takerTokenAmount.div(signedOrder.makerTokenAmount);
                        fillMakerAmount = fillTakerAmount.div(exchangeRate);
                        requiredMakerAmount = isMakerTokenZRX ? fillMakerAmount.plus(signedOrder.makerFee) : fillMakerAmount;
                        if (requiredMakerAmount.greaterThan(makerBalance)) {
                            throw new Error(types_1.ExchangeContractErrs.InsufficientMakerBalance);
                        }
                        if (requiredMakerAmount.greaterThan(makerAllowance)) {
                            throw new Error(types_1.ExchangeContractErrs.InsufficientMakerAllowance);
                        }
                        if (!!isMakerTokenZRX) return [3 /*break*/, 5];
                        return [4 /*yield*/, this.tokenWrapper.getBalanceAsync(zrxTokenAddress, signedOrder.maker)];
                    case 3:
                        makerZRXBalance = _a.sent();
                        isTakerTokenZRX = signedOrder.takerTokenAddress === zrxTokenAddress;
                        if (isTakerTokenZRX) {
                            makerZRXBalance = makerZRXBalance.plus(fillTakerAmount);
                        }
                        return [4 /*yield*/, this.tokenWrapper.getProxyAllowanceAsync(zrxTokenAddress, signedOrder.maker)];
                    case 4:
                        makerZRXAllowance = _a.sent();
                        if (signedOrder.makerFee.greaterThan(makerZRXBalance)) {
                            throw new Error(types_1.ExchangeContractErrs.InsufficientMakerFeeBalance);
                        }
                        if (signedOrder.makerFee.greaterThan(makerZRXAllowance)) {
                            throw new Error(types_1.ExchangeContractErrs.InsufficientMakerFeeAllowance);
                        }
                        _a.label = 5;
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    OrderValidationUtils.prototype.validateFillOrderTakerBalancesAllowancesThrowIfInvalidAsync = function (signedOrder, fillTakerAmount, senderAddress, zrxTokenAddress) {
        return __awaiter(this, void 0, void 0, function () {
            var takerBalance, takerAllowance, isTakerTokenZRX, exchangeRate, fillMakerAmount, requiredTakerAmount, isMakerTokenZRX, takerZRXBalance, takerZRXAllowance;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.tokenWrapper.getBalanceAsync(signedOrder.takerTokenAddress, senderAddress)];
                    case 1:
                        takerBalance = _a.sent();
                        return [4 /*yield*/, this.tokenWrapper.getProxyAllowanceAsync(signedOrder.takerTokenAddress, senderAddress)];
                    case 2:
                        takerAllowance = _a.sent();
                        isTakerTokenZRX = signedOrder.takerTokenAddress === zrxTokenAddress;
                        exchangeRate = signedOrder.takerTokenAmount.div(signedOrder.makerTokenAmount);
                        fillMakerAmount = fillTakerAmount.div(exchangeRate);
                        requiredTakerAmount = isTakerTokenZRX ? fillTakerAmount.plus(signedOrder.takerFee) : fillTakerAmount;
                        if (requiredTakerAmount.greaterThan(takerBalance)) {
                            throw new Error(types_1.ExchangeContractErrs.InsufficientTakerBalance);
                        }
                        if (requiredTakerAmount.greaterThan(takerAllowance)) {
                            throw new Error(types_1.ExchangeContractErrs.InsufficientTakerAllowance);
                        }
                        if (!!isTakerTokenZRX) return [3 /*break*/, 5];
                        isMakerTokenZRX = signedOrder.makerTokenAddress === zrxTokenAddress;
                        return [4 /*yield*/, this.tokenWrapper.getBalanceAsync(zrxTokenAddress, senderAddress)];
                    case 3:
                        takerZRXBalance = _a.sent();
                        if (isMakerTokenZRX) {
                            takerZRXBalance = takerZRXBalance.plus(fillMakerAmount);
                        }
                        return [4 /*yield*/, this.tokenWrapper.getProxyAllowanceAsync(zrxTokenAddress, senderAddress)];
                    case 4:
                        takerZRXAllowance = _a.sent();
                        if (signedOrder.takerFee.greaterThan(takerZRXBalance)) {
                            throw new Error(types_1.ExchangeContractErrs.InsufficientTakerFeeBalance);
                        }
                        if (signedOrder.takerFee.greaterThan(takerZRXAllowance)) {
                            throw new Error(types_1.ExchangeContractErrs.InsufficientTakerFeeAllowance);
                        }
                        _a.label = 5;
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    OrderValidationUtils.prototype.validateRemainingFillAmountNotZeroOrThrow = function (takerTokenAmount, unavailableTakerTokenAmount) {
        if (takerTokenAmount.eq(unavailableTakerTokenAmount)) {
            throw new Error(types_1.ExchangeContractErrs.OrderRemainingFillAmountZero);
        }
    };
    OrderValidationUtils.prototype.validateOrderNotExpiredOrThrow = function (expirationUnixTimestampSec) {
        var currentUnixTimestampSec = utils_1.utils.getCurrentUnixTimestamp();
        if (expirationUnixTimestampSec.lessThan(currentUnixTimestampSec)) {
            throw new Error(types_1.ExchangeContractErrs.OrderFillExpired);
        }
    };
    return OrderValidationUtils;
}());
exports.OrderValidationUtils = OrderValidationUtils;
